### PR TITLE
sysrc: refactor

### LINKED ISSUE #10394: community.general.sysrc fails to parse "sysrc -a" output
### Summary

It seems the sysrc module can't parse the output of `sysrc -a` correctly.
```
MODULE FAILURE: No start of json char found
See stdout/stderr for the exact error


MODULE_STDERR:

Traceback (most recent call last):
  File "<stdin>", line 107, in <module>
  File "<stdin>", line 99, in _ansiballz_main
  File "<stdin>", line 47, in invoke_module
  File "<frozen runpy>", line 226, in run_module
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/tmp/ansible_community.general.sysrc_payload__mjz72wh/ansible_community.general.sysrc_payload.zip/ansible_collections/community/general/plugins/modules/sysrc.py", line 259, in <module>
  File "/tmp/ansible_community.general.sysrc_payload__mjz72wh/ansible_community.general.sysrc_payload.zip/ansible_collections/community/general/plugins/modules/sysrc.py", line 245, in main
  File "/tmp/ansible_community.general.sysrc_payload__mjz72wh/ansible_community.general.sysrc_payload.zip/ansible_collections/community/general/plugins/modules/sysrc.py", line 147, in present
  File "/tmp/ansible_community.general.sysrc_payload__mjz72wh/ansible_community.general.sysrc_payload.zip/ansible_collections/community/general/plugins/modules/sysrc.py", line 133, in exists
ValueError: dictionary update sequence element #16 has length 1; 2 is required
```

The problem seems to be that the `community.general.sysrc` module makes the assumption that all lines that are produced by `sysrc -a` follow the format `key: value`.
However, that's not true.
I have the following in my `/etc/rc.conf`:
```
jail_list="
  ergo
  unifi5
"
```
When running `sysrc -a`, it looks like this:
```
jail_list: 
  ergo
  unifi5
```

### Issue Type

Bug Report

### Component Name

sysrc

### Ansible Version

```console 
ansible [core 2.18.6]
  config file = None
  configured module search path = ['/home/avh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.13/site-packages/ansible
  ansible collection location = /home/avh/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.13.5 (main, Jun 21 2025, 09:35:00) [GCC 15.1.1 20250425] (/usr/bin/python)
  jinja version = 3.1.6
  libyaml = True
```

### Community.general Version

```console
# /usr/lib/python3.13/site-packages/ansible_collections
Collection        Version
----------------- -------
community.general 10.7.1 
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

My Ansible version is 2.18.6.
The host running Ansible is using Python 3.13.5, the FreeBSD target runs 3.11.13.

### Steps to Reproduce

```yaml (paste below)
- name: configure caddy rc script to use custom-build caddy binary
  become: true
    community.general.sysrc:
      state: "present"
      name: "caddy_command"
      value: "/usr/local/bin/custom-caddy"
```

### Expected Results

I expected the sysrc module to configure the key "caddy_command" to the value "/usr/local/bin/custom-caddy".

### Actual Results

The module failed with the error described in the intro

### Code of Conduct

- [x] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by dlundgren**
##### SUMMARY
Refactors the `sysrc` module based on comments from prior PR's. Also fixes #10394.

Closes #10400.

##### ISSUE TYPE
- Refactoring Pull Request

##### COMPONENT NAME
sysrc

##### ADDITIONAL INFORMATION
Refactored the `sysrc` module.

### CODE REVIEW

**Comment by dlundgren** [comment]
I had been sitting on the actual code for a month, and just finished up with the tests this morning. I wanted to get this off my plate before I forgot about it again.

Sorry for a PR that competes with another one.

**Comment by dlundgren** [comment]
~~What are the differences between `FreeBSD 14.2 - 1`, `FreeBSD 14.2 - 2` and `FreeBSD 14.2 - 3`?~~

Never mind, this was my fault on the test for not updating the correct file.

**Comment by dlundgren** [comment]
@vbotka Would you mind checking my logic on the way `load_sysrc_value` works?

I'm fairly certain that using `-c` is the correct answer here so that we don't have to parse the variables. Since we are using `-n self.name` to get the values when appending/subtracting, I figured it would work better to consolidate the value loading to one place.

**Comment by vbotka** [comment]
> using -c is the correct answer

Yes, it is. ``load_sysrc_value`` works fine. This is better than config parsing.

When it is that far, it would be good to understand how sysrc updates rc.conf/rc.conf.d. I'm not sure ``-f`` works in all corner cases. I'll test it.

**Comment by felixfontein** [comment]
One potential problem is that this is a larger refactoring, which needs a very good reason to be backported to `stable-10` as well (by default we only backport new features and refactorings to the current stable branch, which is currently `stable-11`). It is not clear to me how large this refactoring is, and whether it is safe enough to backport to `stable-10` as well. Or should we apply #10400 to `stable-10` and this PR to `main` and `stable-11`?

**Comment by felixfontein** [comment]
(Also please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment) :-) )

**Comment by vbotka** [comment]
> whether it is safe enough

IMO, it is as safe as before.

The error handling is missing (as before). For example, insufficient permissions are silently ignored in ``run_sysrc``

```
-> return (rc, out, err)
(Pdb) p rc
1
(Pdb) p out
''
(Pdb) p err
'/usr/sbin/sysrc: cannot create /etc/rc.conf: Permission denied\n'
```

As a result, the task is reported ``changed``, but the file has no changes.

I would like to proceed on top of this refactoring.

**Comment by vbotka** [comment]
> I'm not sure -f works in all corner cases. I'll test it.

IMO, there are no problems. This PR should be merged as is. The error handling may be added later.

**Comment by dlundgren** [comment]
> The error handling is missing (as before). For example, insufficient permissions are silently ignored in `run_sysrc`
> 
> '/usr/sbin/sysrc: cannot create /etc/rc.conf: Permission denied\n'
> 
> As a result, the task is reported `changed`, but the file has no changes.

Hopefully the permission check I added works to prevent this. 

It did locally, but I have molecule using a vagrant and so I can easily create a test that has `become: false`. Within the Azure pipeline it connects as root, so I can't easily write a test for checking permissions.

**Comment by vbotka** [comment]
> add tests for the use case this PR fixes

There is no reason to test the parsing. After refactoring, the module uses only "sysrc" to parse the configuration files. The ``sysrc`` output is not transformed in the module; hence, testing parsing would mean testing the ``sysrc`` binary, not the module.

**Comment by vbotka** [comment]
I want to propose this patch

```python
@@ -152,7 +152,11 @@ class Sysrc(object):
             return changed(out.split(' -> ')[1].strip().split(self.delim))

     def present(self):
-        if self.load_sysrc_value() == self.value:
+        value = self.load_sysrc_value()
+        if value == self.value:
+            return False
+        if self.value is None:
+            self.value = value
             return False

         if not self.module.check_mode:
@@ -232,6 +236,7 @@ def main():
     try:
         sysrc = Sysrc(module, name, result['value'], result['path'], result['delim'], result['jail'])
         result['changed'] = getattr(sysrc, result['state'])()
+        result['value'] = getattr(sysrc, 'value')
     except OSError as err:
         module.fail_json(msg=str(err))
```
The option ``value`` is not required. If ``value`` is not used, this patch makes the module return the current value if ``state: present``(default). This is how the ``sysrc`` binary behaves. For example,

```console
shell> sysrc k1
k1: v1
```

Without this patch, the module replaces the current value, e.g. ``k1="None"``, if ``name`` is present.

**Comment by felixfontein** [comment]
Instead of `getattr(sysrc, 'value')` please use `sysrc.value` :)

File: changelogs/fragments/10417-sysrc-refactor.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+minor_changes:
+  - sysrc - adjustments to the code (https://github.com/ansible-collections/community.general/pull/10417).
+bugfixes:
+  - sysrc - fixes parsing with multi-line variables (https://github.com/ansible-collections/community.general/issues/10394).
```
Comment (russoz) [inline changelogs/fragments/10417-sysrc-refactor.yml]: Please include the reference to this PR as well:
```suggestion
  - sysrc - fixes parsing with multi-line variables (https://github.com/ansible-collections/community.general/issues/10394, https://github.com/ansible-collections/community.general/pull/10417).
```

File: plugins/modules/sysrc.py [hunk 2]
```diff
@@ -225,32 +220,20 @@ def main():
...
-    delim = module.params.pop('delim')
-    jail = module.params.pop('jail')
     result = dict(
         name=name,
-        state=state,
-        value=value,
-        path=path,
-        delim=delim,
-        jail=jail
+        state=module.params.pop('state'),
```
Comment (russoz) [inline plugins/modules/sysrc.py]: Any particular reason to use `pop` instead of `['state']`? This is my first glance over the code, so please indulge me: it might be obvious by looking at all the code, but right here it _seems_ unnecessary.

File: plugins/modules/sysrc.py [hunk 3]
```diff
@@ -225,32 +220,20 @@ def main():
...
-        rc_value.absent()
-    elif state == 'value_present':
-        rc_value.value_present()
-    elif state == 'value_absent':
-        rc_value.value_absent()
-
-    result['changed'] = rc_value.changed
+    try:
+        sysrc = Sysrc(module, name, result['value'], result['path'], result['delim'], result['jail'])
+        result['changed'] = getattr(sysrc, result['state'])()
```
Comment (russoz) [inline plugins/modules/sysrc.py]: You might want to consider using `StateModuleHelper` to achieve this. See more in https://docs.ansible.com/ansible/latest/collections/community/general/docsite/guide_modulehelper.html

File: tests/integration/targets/sysrc/tasks/main.yml [hunk 4]
```diff
@@ -369,14 +369,66 @@
...
+
+    - name: Get file content
+      shell: "cat /tmp/10394.conf"
+      register: sysrc_10394_content
+
+    - name: Ensure sysrc changed k1 from v1 to v2
+      assert:
+        that:
+          - sysrc_10394_changed.changed
+          - "'k1=\"v2\"' in sysrc_10394_content.stdout_lines"
```
Comment (russoz) [inline tests/integration/targets/sysrc/tasks/main.yml]: I think the readability would be improved by:
```suggestion
          - >
            'k1="v2"' in sysrc_10394_content.stdout_lines
```

File: plugins/modules/sysrc.py [hunk 5]
```diff
@@ -201,6 +194,8 @@ def run_sysrc(self, *args):
         cmd.extend(args)
 
         (rc, out, err) = self.module.run_command(cmd)
+        if err.find("Permission denied") > 0:
```
Comment (russoz) [inline plugins/modules/sysrc.py]: This works, of course, but it is more commonly expressed as:
```suggestion
        if "Permission denied" in err:
```

File: plugins/modules/sysrc.py (hunk 3 above)
Comment (dlundgren) [reply plugins/modules/sysrc.py]: Thanks for pointing me to that.

This is much nicer than managing this manually.

File: plugins/modules/sysrc.py (hunk 2 above)
Comment (dlundgren) [reply plugins/modules/sysrc.py]: It's what I originally wrote based on other modules from many, many years ago. Switching to the `StateModuleHelper` as in your other suggestion will resolve this.

**Comment by vbotka** [comment]
Without the name validity test, the error ``name contains characters not allowed in shell`` is silently ignored. As a result, the module reports ``changed``, but the file has no changes.

**Comment by dlundgren** [comment]
@vbotka thanks for finding that! I added the check back in and a test for it as well.

**Comment by vbotka** [comment]
The below task is not idempotent. It always reports ``changed``

```yaml
    - community.general.sysrc:
        name: static_routes
        state: absent
```

``static_routes`` is declared in defaults

```console
shell> grep -r "^static_routes" /etc/defaults
/etc/defaults/rc.conf:static_routes=""		# Set to static route list (or leave empty).
```

Therefor , ``_get`` finds it

```console
shell> sysrc -c static_routes
shell> echo $?
0
```

Then, ``sysrc -x static_routes`` is executed. It always returns ``rc=0`` with empty ``out`` and ``err``. From the output of the command, it is not possible to recognize whether anything was changed or not

```python
167  	        (rc, out, err) = self.module.run_command(cmd)
168  ->	        if "Permission denied" in err:
169  	            raise OSError(errno.EACCES, "Permission denied for %s" % self.vars.path)
170  	
171  	        return rc, out, err
172  	
173  	    def state_absent(self):
(Pdb) p rc
0
(Pdb) p out
''
(Pdb) p err
''
```

Then, the task is set ``changed = True``

```python
181  ->	        self.changed = True
```

**Comment by vbotka** [comment]
Quoting man sysrc:

```text
-c	Check  only.   For  querying,  return success if all requested
	variables are set (even	if NULL), otherwise return error  sta-
	tus.   For  assignments,  return success if no changes are re-
	quired,	otherwise failure.  If verbose	(see  "-v")  prints  a
	message	 stating  whether variables are	set and/or changes are

-v	Verbose.   Print  the pathname of the specific rc.conf(5) file
	where the directive was	found.
```
If ``static_routes=""`` is in ``/etc/rc.conf``, the below command tells it

```console
shell> sysrc -c -v static_routes="foo, bar"
/etc/rc.conf: static_routes: would change from `' to `foo, bar'
```
It also tells that a ``name`` is in the ``defaults`` only

```console
shell> sysrc -c -v static_routes="foo, bar"
/etc/defaults/rc.conf: static_routes: would change from `' to `foo, bar'
```

The patch below makes the module return ``self.changed = False`` when the ``name`` is in the ``defaults`` only, e.g. no deletion is needed.

```python
diff --git a/plugins/modules/sysrc.py b/plugins/modules/sysrc.py
index 0b57becb..223c0463 100644
--- a/plugins/modules/sysrc.py
+++ b/plugins/modules/sysrc.py
@@ -161,6 +161,20 @@ class Sysrc(StateModuleHelper):

         return False

+    def _sysrc_defaults(self):
+        """ Return True, if name is in /etc/defaults/rc.conf only. """
+
+        cmd = [self.sysrc]
+        if self.vars.jail:
+            cmd += ['-j', self.vars.jail]
+        cmd += ['-c', '-v', '%s="foo, bar"' % self.vars.name]
+
+        (rc, out, err) = self.module.run_command(cmd)
+        if "/etc/defaults/rc.conf" in out.split(':', 1):
+            return True
+
+        return False
+
     def _sysrc(self, *args):
         cmd = [self.sysrc, '-f', self.vars.path]
         if self.vars.jail:
@@ -178,9 +192,14 @@ class Sysrc(StateModuleHelper):
             return

         if not self.check_mode:
-            self._sysrc('-x', self.vars.name)

-        self.changed = True
+            if self._sysrc_defaults():
+                self.changed = False
+                return
+
+            self._sysrc('-x', self.vars.name)
+            self.changed = True
+            return

     def state_present(self):
         value = self._get()
```

File: plugins/modules/sysrc.py [hunk 6]
```diff
@@ -102,157 +103,120 @@
...
+            name=dict(type='str', required=True),
+            value=dict(type='str', default=None),
+            state=dict(type='str', default='present', choices=['absent', 'present', 'value_present', 'value_absent']),
+            path=dict(type='str', default='/etc/rc.conf'),
+            delim=dict(type='str', default=' '),
+            jail=dict(type='str', default=None)
+        ),
+        supports_check_mode=True
+    )
+    output_params = ('value')
```
Comment (russoz) [inline plugins/modules/sysrc.py]: This evaluates to the string, not to a tuple. You might want to adjust it to

```suggestion
    output_params = ('value',) 
```

File: plugins/modules/sysrc.py [hunk 7]
```diff
@@ -102,157 +103,120 @@
...
+    use_old_vardict = False
 
-        if not self.module.check_mode:
-            (rc, out, err) = self.run_sysrc("%s=%s" % (self.name, self.value))
+    def __init_module__(self):
+        # OID style names are not supported
+        if not re.match('^[a-zA-Z0-9_]+$', self.vars.name):
```
Comment (russoz) [inline plugins/modules/sysrc.py]: Your regexp works, but that can be expressed as:

```suggestion
        if not re.match('\w+', self.vars.name):
```
`\w` being a convenience shortcut for those characters, and `^` and `$` being implied in`re.match`, unlike `re.search`.

File: plugins/modules/sysrc.py [hunk 8]
```diff
@@ -102,157 +103,120 @@
...
+            return None
 
-        setstring = '%s+=%s%s' % (self.name, self.delim, self.value)
-        (rc, out, err) = self.run_sysrc(setstring)
-        if out.find("%s:" % self.name) == 0:
-            values = out.split(' -> ')[1].strip().split(self.delim)
-            if self.value in values:
-                self.changed = True
+        (rc, out, err) = self._sysrc('-n', self.vars.name)
+        if "unknown variable" in err or "unknown variable" in out:
```
Comment (russoz) [inline plugins/modules/sysrc.py:148]: Why do we need to test in both? I would suppose the command always send that specific error message to one of them.

**Review by russoz — COMMENTED** [review]
Hi @dlundgren thanks for your continued efforts.

Couple of additional comments here.

File: plugins/modules/sysrc.py (hunk 7 above)
Comment (dlundgren) [reply plugins/modules/sysrc.py]: After testing this, I still needed `r'^\w+$'` to ensure that a `not.valid.var` would fail, otherwise it was matching on `not` for me, and the test for detecting invalid variables was not failing as expected.

File: plugins/modules/sysrc.py (hunk 8 above)
Comment (dlundgren) [reply plugins/modules/sysrc.py:148]: `sysrc` prior to FreeBSD 11.2 wrote the `unknown variable` warning to `stdout` and not `stderr`. [FreeBSD bug report](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=229806).

I added a comment reflecting this.

**Comment by dlundgren** [comment]
@vbotka My idea behind using `sysrc -c` was to try to avoid the defaults file, I'm not sure why the tests passed originally, they should have failed.

I borrowed your idea, and now supply `-v` with a check to verify the value is coming from the file that's been specified.

**Comment by felixfontein** [comment]
Looks like CI now passes (https://dev.azure.com/ansible/community.general/_build/results?buildId=152570&view=results), but GitHub shows the previous CI run...

File: plugins/modules/sysrc.py (hunk 7 above)
Comment (russoz) [reply plugins/modules/sysrc.py]: I was puzzled by this so I checked the docs:

https://docs.python.org/3/library/re.html#re.match

Apparently `^` is indeed implied in `match()`, but `$` is not. So this should work as '\w+$'. But this is hair-splitting now, no need to change it.

File: tests/integration/targets/sysrc/tasks/main.yml [hunk 9]
```diff
@@ -369,14 +369,79 @@
...
+    ##
+    - name: Ensure failure on OID style name since sysrc does not support them
+      sysrc:
+        name: not.valid.var
+        value: test
+      register: sysrc_name_check
+      failed_when:
+        - sysrc_name_check is not failed
+        - >
+          'Name may only contain alpha-numeric and underscore characters' != sysrc_name_check.msg
```
Comment (russoz) [inline tests/integration/targets/sysrc/tasks/main.yml:435]: Nit pick.

We do not gain much by breaking this into the other line.
```suggestion
        - 'Name may only contain alpha-numeric and underscore characters' != sysrc_name_check.msg
```

Feel free to disregard it for this PR, though.

File: plugins/modules/sysrc.py (hunk 8 above)
Comment (russoz) [reply plugins/modules/sysrc.py:148]: Just keep in mind that we (specially you) do not want to support outdated systems forever. It would be interesting to draw a line - not here, not in this PR - and say we support FreeBSD X.Y or greater. Once you have that, you can start making plans to stop ~~dragging this bag of bricks~~ supporting those features.

File: tests/integration/targets/sysrc/tasks/main.yml (hunk 9 above)
Comment (felixfontein) [reply tests/integration/targets/sysrc/tasks/main.yml:435]: This line would be invalid YAML, the `>` is needed to make YAML not interpret the `'`.

**Comment by felixfontein** [comment]
If nobody objects I'll merge this later today for the 11.1.1 release.

File: plugins/modules/sysrc.py (hunk 8 above)
Comment (felixfontein) [reply plugins/modules/sysrc.py:148]: Deprecating support can happen in any feature release, but actually dropping support can only happen in a major release some time after deprecation.

**Comment by felixfontein** [comment]
@dlundgren thanks for improving the module!
@vbotka @russoz thanks for reviewing and testing!

**Comment by jlduran** [comment]
Hello!

I believe this refactoring broke the ability to discover rc variables configured under `/etc/defaults/vendor.conf` (I noticed `sysrc -e` was removed).

As a consequence, the `sysrc` module now tries to add an rc variable to `/etc/rc.conf` without realizing it is already defined under `/etc/defaults/vendor.conf`.

Thank you!

**Comment by jlduran** [comment]
> Hello!
> 
> I believe this refactoring broke the ability to discover rc variables configured under `/etc/defaults/vendor.conf` (I noticed `sysrc -e` was removed).
> 
> As a consequence, the `sysrc` module now tries to add an rc variable to `/etc/rc.conf` without realizing it is already defined under `/etc/defaults/vendor.conf`.
> 
> Thank you!

Sorry for the noise. I realize now this was a design decision, and we are supposed to explicitly specify: `path: /etc/defaults/vendor.conf`, for example.
